### ISSUE TITLE
Is there an explicit way of clearing cache?

### ISSUE THREAD

**Issue by zhiyuanGH** [containerd/stargz-snapshotter#1213]
I want to explore how can parameters affect the performance of lazy-pulling, e.g. ChunkSize. This is because some recent research papers report that lazy-pulling is ineffective when network latency is high, since the multiple http requests caused by on-demand fetching suffer from high latency. 

I try to repeat running a same container to collect performance metrics. I remove images in nerdctl and restart stargz-snapshotter after each run, however it still logs hit layer cache:
Apr 24 16:44:35 gh-virtual-machine containerd-stargz-grpc[274521]: {"key":"default/66/extract-491987201-d0tB sha256:2414385fd51d34e07d564ec6041ee66de902424f028528bce52743d92b1bc875","level":"debug","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/66/fs","msg":"hit layer cache \"ghcr.io/stargz-containers/node:17.8.0-esgz/sha256:e0912037615853fecb2a3de7eacb11e30debbc7ae24d64a7ad69bf9fa84ea10a\"","parent":"sha256:a3926353a4b2389bed133fe4b9f8bdb8439529ba6a965b37ef0c1a7921043a00","src":"ghcr.io/stargz-containers/node:17.8.0-esgz/sha256:e0912037615853fecb2a3de7eacb11e30debbc7ae24d64a7ad69bf9fa84ea10a","time":"2023-04-24T16:44:35.493692205+08:00"}

I've notice that #309  implements some cache clearing functions when unmount.

Is there any explicit way to clear cache thoroughly? I think i shouldn't profile the system with cache un-cleared.

**Comment by zhiyuanGH** [comment]
Perhaps I misunderstand the log. In #600  I see: " layer are cached When filesystem.Mount is called for the first layer in an image, stargz-snapshotter resolves all layers in that image in parallel and caches these resolved layers metadata in LRU cache. When filesystem.Mount is called for the neighbouring layers of that image, the cached layer can be used for speeding up the mounts."

So that hit layer cache will be a common phenomenon even when run a container for the first time.

If this is correct, could I change my question to: Is their any factors will affect repeating profiling the stargz-snapshotter? Can I just delete the image and repeat profiling stargz on the same image?

**Comment by ktock** [comment]
Yes, stargz snapshotter caches resources for a while for future use.
If you want to ensure to clean up resources on the node, you can stop `containerd-stargz-grpc` process, remove files under stargz-snapshotter's root dir (typically `/var/lib/containerd-stargz-grpc/`) and restart stargz-snapshotter.

**Comment by ktock** [comment]
Yes, landmark is respected than PrefetchSize and no configuration flag is provided for limiting the prefetching size. Maybe we can introduce `MaxPrefetchSize` or `PrefetchSizeLimit` for this purpose. Patches are welcome if you're interested in.

**Comment by sevetseh28** [comment]
hey @ktock  - I've seen that restarting my containers cause a constant increase on disk , even though I see `hit layer cache` logs in the stargz-snapshotter.service it seems that it's redownloading the layers or not using the cache.
I also noticed these two directories being used:
`/var/lib/docker/rootfs/stargz`
`/var/lib/containerd-stargz-grpc`

Any ideas? Maybe some misconfiguration?

**Comment by ktock** [comment]
> I've seen that restarting my container that was lazily pulled cause a constant increase on disk , even though I see hit layer cache logs in the stargz-snapshotter.service it seems that it's redownloading the layers or not using the cache.

What cli or orchestrator are you using (docker,  nerdctl, k8s)? It would be great if there is a reproducer.

**Comment by sevetseh28** [comment]
> > I've seen that restarting my container that was lazily pulled cause a constant increase on disk , even though I see hit layer cache logs in the stargz-snapshotter.service it seems that it's redownloading the layers or not using the cache.
> 
> What cli or orchestrator are you using (docker, nerdctl, k8s)? It would be great if there is a reproducer.

Im using `docker` as a non-root user (`ec2-user` in Amazon Linux 2023). These are my configs:

docker's daemon.json:
```json
{
  "features": {
    "containerd-snapshotter": true
  },
  "storage-driver": "stargz"
}
```

containerd's config.toml:
```toml
version = 2

[proxy_plugins]
  [proxy_plugins.stargz]
    type = "snapshot"
    address = "/run/containerd-stargz-grpc/containerd-stargz-grpc.sock"
```

stargz-snapshotter.service:

```
[Unit]
Description=stargz snapshotter
After=network.target
Before=containerd.service

[Service]
ExecStartPost=/bin/sh -c 'chown root:docker /run/containerd-stargz-grpc/containerd-stargz-grpc.sock && chmod 660 /run/containerd-stargz-grpc/containerd-stargz-grpc.sock'
Environment="DOCKER_CONFIG=/home/ec2-user/.docker/"
Type=notify
Environment=HOME=/root
ExecStart=/usr/local/bin/containerd-stargz-grpc --log-level=debug --config=/etc/containerd-stargz-grpc/config.toml
Restart=always
RestartSec=1

[Install]
WantedBy=multi-user.target
```

File permissions:
![Image](https://github.com/user-attachments/assets/b027857a-0be3-43fe-abab-4d364577e24c)

Notice constant increase on `fscache` and `httpcache` on container restart:
![Image](https://github.com/user-attachments/assets/e667e27b-02e1-4b70-ba4a-01160e068433)

Thank you!

**Comment by sevetseh28** [comment]
First I see many 401 Unauthorized because im using ECR - but I have setup a credential helper under my docker config and it successfully attempts to reauthenticate. Then I see some prefetch and complete to prefetch logs, and many hit layer cache logs. Some examples:

```
{"level":"info","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/44/fs","msg":"Received status code: 401 Unauthorized. Refreshing creds...","src":"<account-id-masked>.dkr.ecr.us-east-1.amazonaws.com/<repo-id-masked>/sha256:980eb8e44f332aa233462b9cf9afe235ffb8c334cb07e825719ef0456d4d1a81","time":"2025-04-08T08:33:18.109952403Z"}
...
...
{"layer_sha":"sha256:cc2813e28e0dc08b9c699f50a0199468c6cc16a894fbecfb36113136abef7309","level":"info","metrics":"latency","msg":"value=0.013161 milliseconds","operation":"prefetch_download","time":"2025-04-08T08:33:19.051287435Z"}
{"layer_sha":"sha256:cc2813e28e0dc08b9c699f50a0199468c6cc16a894fbecfb36113136abef7309","level":"info","metrics":"latency","msg":"value=4.412837 milliseconds; prefetch_size=2603 bytes","operation":"prefetch_total","time":"2025-04-08T08:33:19.055682878Z"}
{"level":"debug","msg":"completed to prefetch","time":"2025-04-08T08:33:19.055689236Z"}
{"layer_sha":"sha256:cc2813e28e0dc08b9c699f50a0199468c6cc16a894fbecfb36113136abef7309","level":"info","metrics":"latency","msg":"value=4.300907 milliseconds","operation":"prefetch_decompress","time":"2025-04-08T08:33:19.055653824Z"}
...
...

{"level":"debug","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/44/fs","msg":"verified","time":"2025-04-08T08:33:19.086834974Z"}
{"level":"info","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/44/fs","msg":"fusermount detected","time":"2025-04-08T08:33:19.086926024Z"}
...
...
{"level":"info","msg":"preparing filesystem mount at mountpoint=/var/lib/containerd-stargz-grpc/snapshotter/snapshots/41/fs","time":"2025-04-08T08:33:19.124661481Z"}
...

{"level":"debug","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/41/fs","msg":"hit layer cache \"<hidden>.dkr.ecr.us-east-1.amazonaws.com/<hidden>/sha256:9b710ae23bfeeffcc6a9df95af62a6cf6327e55844d22a819529a02e643b55e1\"","src":"<hidden>.dkr.ecr.us-east-1.amazonaws.com/<hidden>/sha256:9b710ae23bfeeffcc6a9df95af62a6cf6327e55844d22a819529a02e643b55e1","time":"2025-04-08T08:33:19.433611580Z"}
...

```

**Comment by sevetseh28** [comment]
I think it shouldn't even try to contact ECR the second time. The credentials are valid for 12 hours, yet, I see the 401 trying to contact ECR again only after an hour of the initial lazy pull. All layers should've been downloaded and cached.

**Comment by sevetseh28** [comment]
hey @ktock - I also found this issue https://github.com/containerd/stargz-snapshotter/issues/1349
I saw that the issue there happens with k8s but maybe it's somehow related

**Comment by sevetseh28** [comment]
doing `ctr-remote snapshot --snapshotter=stargz ls` also lists nothing, even though the lazy pulling is working and the Snapshots exist on disk.

**Comment by sevetseh28** [comment]
using docker CLI -> `docker images` also doesn't list the image

**Comment by sevetseh28** [comment]
@ktock  more debug insights:
I filtered the logs based on snapshot 35 and `preparing filesystem mount at...` - It seems it's remounting after every reboot, so it's probably accumulating on `fscache` even though I still see `hit layer cache` logs
![Image](https://github.com/user-attachments/assets/98665ad2-4b42-4a8f-a3e4-c021a1951bda)
